Source/quests.cpp

```diff
@@ -378,137 +378,118 @@ void DrawSkelKing(int q, int x, int y)
 
 void DrawWarLord(int x, int y)
 {
-	int rw, rh;
-	int i, j;
-	BYTE *sp;
-	int v;
-
-	auto setp = LoadFileInMem("Levels\\L4Data\\Warlord2.DUN");
-	rw = setp[0];
-	sp = &setp[2];
-	rh = *sp;
-	sp += 2;
-	setpc_w = rw;
-	setpc_h = rh;
+	auto dunData = LoadFileInMem<uint16_t>("Levels\\L4Data\\Warlord2.DUN");
+
+	int width = SDL_SwapLE16(dunData[0]);
+	int height = SDL_SwapLE16(dunData[1]);
+
 	setpc_x = x;
 	setpc_y = y;
-	for (j = y; j < y + rh; j++) {
-		for (i = x; i < x + rw; i++) {
-			if (*sp != 0) {
-				v = *sp;
-			} else {
-				v = 6;
-			}
-			dungeon[i][j] = v;
-			sp += 2;
+	setpc_w = width;
+	setpc_h = height;
+
+	uint16_t *tileLayer = &dunData[2];
+
+	for (int j = y; j < y + height; j++) {
+		for (int i = x; i < x + width; i++) {
+			uint8_t tileId = SDL_SwapLE16(tileLayer[j * width + i]);

```

Comment:
There's a huge number of SDL_SwapLE16 calls, which return a UInt16, which then get assigned to uint8_t.

Looking at the old code, this should work, but if you raise the warning level those will all spam type narrowing warnings like crazy.

Since this seems to be such a common operation, maybe add a function or macro that does the 'read uint16 -> swap endian -> cast to uint8' in one step, and add a comment why this is valid and needed. tile_cast or something similar.

Reply:
This is a bit of a general issue, so I would rather that it be dealt with at another point.

Also instead of the proposal here I would be more inclined to changing the internal size to match the format, it would use about 3KiB more ram but we would have support for much later tilsets and maps.